File: csharp/1477.find-two-non-overlapping-sub-arrays-each-with-target-sum.cs
/*
 * @lc app=leetcode id=1477 lang=csharp
 *
 * [1477] Find Two Non-overlapping Sub-arrays Each With Target Sum
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int MinSumOfLengths(int[] arr, int target) {
        var map = new Dictionary<int, int>();
        int acc = 0, n = arr.Length, ret = int.MaxValue, size = int.MaxValue;
        map[0] = -1;
        for (int i = 0; i < n; ++i) {
            acc += arr[i];
            map[acc] = i;
        }
        acc = 0;
        for (int i = 0; i < n; ++i) {
            acc += arr[i];
            if (map.TryGetValue(acc - target, out int j)) {
                size = Math.Min(size, i - j);
            }
            if (size < int.MaxValue && map.TryGetValue(acc + target, out int k)) {
                ret = Math.Min(size + k - i, ret);
            }
        }
        return ret == int.MaxValue ? -1 : ret;
    }
}
// @lc code=end

File: java/1477.find-two-non-overlapping-sub-arrays-each-with-target-sum.java
/*
 * @lc app=leetcode id=1477 lang=java
 *
 * [1477] Find Two Non-overlapping Sub-arrays Each With Target Sum
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start

class Solution {
    public int minSumOfLengths(int[] arr, int target) {
        int n = arr.length, acc = 0, minSize = n + 1, ret = n + 1;
        Map<Integer, Integer> valToIndex = new HashMap<>(n + 1);
        valToIndex.put(0, -1);
        for (int i = 0; i < n; ++i) {
            acc += arr[i];
            valToIndex.put(acc, i);
        }
        acc = 0;
        for (int i = 0; i < n; ++i) {
            acc += arr[i];
            Integer pi = valToIndex.get(acc - target);
            if (pi != null) {
                minSize = Math.min(minSize, i - pi);
            }
            if (minSize < n) {
                Integer ni = valToIndex.get(acc + target);
                if (ni != null) {
                    ret = Math.min(ret, minSize + ni - i);
                }
            }
        }
        return ret <= n ? ret : -1;
    }
}
// @lc code=end

